### ISSUE TITLE
The improvement of GPU execution efficiency encounters a bottleneck ？ 

### ISSUE THREAD

**Issue by YeahNew** [NVIDIA/spark-rapids#4877]
Hello,
    I recently used spark3.2 +GPU to do a test based on the TPC-DS dataset, and the entire TPC-DS data scale is 1TB(located in HDFS). But I encountered a problem that I couldn't understand, and I hope to get your help.
    The SQL statement tested is as follows:
    select count(*) from (select c_last_name from store_sales, customer where store_sales.ss_customer_sk = customer.c_customer_sk) hot_cust limit 100;

Three tables are used here, and their sizes are: store_sales(387.97GB), date_dim(9.84GB),customer(1.52GB)

When submitting the application, the driver and executor parameters are set as follows:
DRIVER_OPTIONS="--master spark://10.42.101.70:7078 --jars ${CUDF_JAR},${RAPIDS_JAR} \ 
--conf spark.executor.extraClassPath=${CUDF_JAR},${RAPIDS_JAR} \
--conf spark.driver.extraClassPath=${CUDF_JAR},${RAPIDS_JAR} \
--conf spark.plugins=com.nvidia.spark.SQLPlugin \
--conf spark.rapids.sql.incompatibleOps.enabled=true \
--conf spark.rapids.memory.gpu.pooling.enabled=true \
--conf spark.shuffle.service.enabled=false \
--conf spark.rapids.shuffle.transport.enabled=true \
--conf spark.rapids.shuffle.manager.enabled=true \
--conf spark.executor.resource.gpu.amount=1 \
--conf spark.rapids.memory.gpu.allocFraction=0.9 \
--conf spark.rapids.shuffle.transport.enabled=true \
--conf spark.rapids.sql.enabled=true \
--conf spark.rapids.sql.concurrentGpuTasks=4 \
--conf spark.task.resource.gpu.amount=0.125 \
--conf spark.driver.maxResultSize=60g \
--conf spark.locality.wait=0s \
--conf spark.rapids.sql.hasNans=true \
--conf spark.executor.extraJavaOptions="-Dai.rapids.cudf.prefer-pinned=true" \ 
--total-executor-cores 24 \
--driver-memory 50g \
--conf spark.default.parallelism=30 \
--conf spark.sql.adaptive.enabled=false \
--driver-java-options -Dlog4j.configuration=file:///${OUTPUT_DIR}/log4j.properties"

EXECUTOR_OPTIONS="--executor-cores 8 \
--conf spark.executor.memory=50g \
--conf spark.task.cpus=1 \
--conf spark.sql.shuffle.partitions=30 \
--conf spark.executor.extraJavaOptions=-Dlog4j.configuration=file:///${OUTPUT_DIR}/log4j.properties \
--conf spark.sql.crossJoin.enabled=true"

The Spark cluster is built on a physical node with only one worker, and the submitted tasks are in StandAlone mode. A total of 56 cores, 376GB of memory. I kept executor-num=3 unchanged (that is, the total executors remained unchanged), and adjusted executor-cores=2, 4, 8, 12,16. The corresponding application time was: 78s, 57s, 46s, 49s,50s. It was found that the total- executor-cores >=24, the time-consuming is almost unchanged. 

I monitored CPU utilization, IO, and found no bottlenecks. Then I analyzed the task avg time in the longest stage of each application in the spark web UI interface(The longest time-consuming stage in the four test cases are stage2, and stage3). The average time of tasks has been increasing .As shown in the table.
![图片](https://user-images.githubusercontent.com/33194373/155986351-d3ea73e3-e3ba-438f-a8b7-a12bbae35b35.png)
From the table results, when the number of cpu cores increases to a certain level, the execution efficiency hardly increases, and there will be a decline. What is the reason? 
Under the same number of cores, it is found that the execution time of using GPU is about half of that of using CPU (other parameters remain unchanged, just set whether to start GPU). With GPU-accelerated sql, the execution performance doesn't seem to improve significantly（total-executor-cores=48,  CPU:96.6s----->GPU:48.3s).
Is it related to file size and data distribution? 

stage2 picture:
![图片](https://user-images.githubusercontent.com/33194373/155988928-61f41057-0655-404f-9128-30a702754fb7.png)
stage3 picture:
![图片](https://user-images.githubusercontent.com/33194373/155989023-985e6992-a19f-4f5c-baf5-6c707c8e5605.png)

**Comment by revans2** [comment]
I don't know how much background you have in this so please excuse me if I explain some things that you already know. 

TLDR; It looks like your GPU is fully utilized and that is the bottleneck. Would need to look at GPU stats to know that for sure.

You have set your GPU concurrent task parallelism to 4.  This is the number of tasks that can share a GPU at any one point in time.  We limit the number of tasks that can be on the GPU so we don't use up all of the GPU's memory, but we have also seen some issues with compute on some GPUs with lots of memory and I should update the docs to reflect this. We keep it separate from the parallelism of the CPU, because it allows a user to configure the query to run with more CPU cores. This is especially helpful if there are portions of the query that require a lot of CPU. Adding more CPU cores beyond that parallelism number only impacts the parallelism of what can be run on the CPU. Looking at your screen shots it is clear that all of the query is running on the GPU. This means that once you get above 4 tasks the extra tasks may be able to do more I/O in parallel, but will not improve the computation speed.  In many cases we have seen this be a big win for performance, especially if the I/O is relatively slow. The reason why the average task time goes up for 5 tasks over 4 tasks is because Spark is measuring that from the perspective of the CPU, so if one of the tasks is blocked on a semaphore waiting to get on the GPU, then it looks like that task took longer to complete.

In your case there could be a number of different things happening and I probably would need more metrics to be 100% sure what it is. One of the first things that comes to mind is HDFS limitations. I don't think this is the case here, but you might want to check your network utilization and also the disk I/O on your HDFS nodes. You could also look at the buffer time metrics for reading the data. This should really only impact the first stage, and it looks like it is scaling much better than the second stage. So, like I said, it is not likely the problem.

The second stage appears to be where a lot of the issue is. The average task time appears to grow with the number of tasks put on it, and really jumps up with 4 tasks. The GPU is great and is able to do sub-linear scaling so long as it has enough compute/memory bandwidth to be able to keep up. After that point it has to try to schedule what to run next. In our testing we have seen it struggle when there are a lot more things to do than resources to fulfill those requests. 

When running with just one task we tend to see a number of gaps in GPU utilization. This is because of I/O or just the time it takes for JNI to call down to the GPU to issue the next set of commands to run per task.  With 2 tasks we often see these gaps really fill in. One task will be on the GPU while the other is getting something else queued up to run. This is why the biggest gains often come with setting a parallelism of 2 (a 37% improvement in your case). After that there are diminishing returns, and it really depends on how good your hardware is, and the algorithms that you are going to run on it.

For your setup I would recommend trying to set your gpuParallelism to 3, and the number of tasks that you have to 6.  Ideally that would allow half of the tasks to be reading data while the other half are doing computation and you can overlap as much of that as possible.  In your setup that is not as likely to make as much of a difference because most of your I/O is local, and because of the amount of RAM you have on your device it is likely all coming from the page cache.

**Comment by YeahNew** [comment]
> I don't know how much background you have in this so please excuse me if I explain some things that you already know.
> 
> TLDR; It looks like your GPU is fully utilized and that is the bottleneck. Would need to look at GPU stats to know that for sure.
> 
> You have set your GPU concurrent task parallelism to 4. This is the number of tasks that can share a GPU at any one point in time. We limit the number of tasks that can be on the GPU so we don't use up all of the GPU's memory, but we have also seen some issues with compute on some GPUs with lots of memory and I should update the docs to reflect this. We keep it separate from the parallelism of the CPU, because it allows a user to configure the query to run with more CPU cores. This is especially helpful if there are portions of the query that require a lot of CPU. Adding more CPU cores beyond that parallelism number only impacts the parallelism of what can be run on the CPU. Looking at your screen shots it is clear that all of the query is running on the GPU. This means that once you get above 4 tasks the extra tasks may be able to do more I/O in parallel, but will not improve the computation speed. In many cases we have seen this be a big win for performance, especially if the I/O is relatively slow. The reason why the average task time goes up for 5 tasks over 4 tasks is because Spark is measuring that from the perspective of the CPU, so if one of the tasks is blocked on a semaphore waiting to get on the GPU, then it looks like that task took longer to complete.
> 
> In your case there could be a number of different things happening and I probably would need more metrics to be 100% sure what it is. One of the first things that comes to mind is HDFS limitations. I don't think this is the case here, but you might want to check your network utilization and also the disk I/O on your HDFS nodes. You could also look at the buffer time metrics for reading the data. This should really only impact the first stage, and it looks like it is scaling much better than the second stage. So, like I said, it is not likely the problem.
> 
> The second stage appears to be where a lot of the issue is. The average task time appears to grow with the number of tasks put on it, and really jumps up with 4 tasks. The GPU is great and is able to do sub-linear scaling so long as it has enough compute/memory bandwidth to be able to keep up. After that point it has to try to schedule what to run next. In our testing we have seen it struggle when there are a lot more things to do than resources to fulfill those requests.
> 
> When running with just one task we tend to see a number of gaps in GPU utilization. This is because of I/O or just the time it takes for JNI to call down to the GPU to issue the next set of commands to run per task. With 2 tasks we often see these gaps really fill in. One task will be on the GPU while the other is getting something else queued up to run. This is why the biggest gains often come with setting a parallelism of 2 (a 37% improvement in your case). After that there are diminishing returns, and it really depends on how good your hardware is, and the algorithms that you are going to run on it.
> 
> For your setup I would recommend trying to set your gpuParallelism to 3, and the number of tasks that you have to 6. Ideally that would allow half of the tasks to be reading data while the other half are doing computation and you can overlap as much of that as possible. In your setup that is not as likely to make as much of a difference because most of your I/O is local, and because of the amount of RAM you have on your device it is likely all coming from the page cache.

@revans2 
Thank you for your reply，
You mentioned above that you may need to get more metrics to to be 100% sure what it is, can you give some examples? I would like to show you. 

"The second stage appears to be where a lot of the issue is. The average task time appears to grow with the number of tasks put on it, and really jumps up with 4 tasks" you mentioned above. Sorry, here I don't fully understand the explanation. Is the diminishing returns because of the GPU itself or the algorithm? 

I have tried to set  gpuParallelism to 3, and the number of tasks to 6, The execution efficiency will drop further, which doesn't seem to be a good setup for my case.

**Comment by revans2** [comment]
Also a few more questions. From the size it looks like you are reading from the original data format, "|" separated. I assume you are reading the data as floats and not decimal values?

I should also add that I would be interested in seeing the number of tasks run, simply because you are not setting spark.sql.files.maxPartitionBytes, which will default to 128MB, which is rather small for the GPU. But I first want to match your execution before I start making recommendations to change, because my first assumption was way off.
